lib/mpl_toolkits/mplot3d/art3d.py

```diff
@@ -455,14 +455,24 @@ def do_3d_projection(self, renderer):
 
         fcs = (zalpha(self._facecolor3d, vzs) if self._depthshade else
                self._facecolor3d)
-        fcs = mcolors.to_rgba_array(fcs, self._alpha)
-        self.set_facecolors(fcs)
-
         ecs = (zalpha(self._edgecolor3d, vzs) if self._depthshade else
                self._edgecolor3d)
+
+        # Sort the points based on z coordinates
+        # Put vzs first to sort based on z value
+        z_markers = sorted(zip(vzs, np.column_stack([vxs, vys]), fcs, ecs),
```

Comment:
I'm a bit worried that this could slow down the rendering when using many patches.

I know it's bad practice to base such an assumption on pure gut feeling. But before I'm going into any tests. Have you done benchmarking on the performance impact?

Reply:
Yes I had ran some benchmarking beforehand. I didn't notice any noticeable impact on performance.

I used the following script
```python
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

import random as rand
SIZE=10000
x = np.array([rand.uniform(-50, 50) for i in range(SIZE)])
y = np.array([rand.uniform(-50, 50) for i in range(SIZE)])
z = np.array([rand.uniform(-50, 50) for i in range(SIZE)])
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
patches = ax.scatter(x, y, z, s=100, c=x)

plt.savefig("tmp.png")
```
Ran the script 5 times both before and after fix and didn't find any noticeable differences.
[perf_fix.log](https://github.com/matplotlib/matplotlib/files/1835558/perf_fix.log)
[perf_vanilla.log](https://github.com/matplotlib/matplotlib/files/1835559/perf_vanilla.log)

Reply:
These timings would have been swamped by all of the imports and figure creation and such. A real benchmark would involve seeing how long it would take to set up the plot, and then have it rotate around, like in one of the animation examples.

Reply:
I did a quick check on this part of the code, with numbers of points increased to 1000 000 (I think 100x100x100 is not unrealistic). With that the present sorting/zipping implementation is in the 100ms regime.

That's a size which gets relevant for performance in interactive use.

In a quick test, it's possible to get a roughly 10x speedup by using `numpy.argsort` on vzs and applying the index array to the other quantities.